### PR TITLE
Flux Validation

### PR DESCRIPTION
**Pull request by wesleytruong**
# This pr implements the validator class for flux following the method discussed in Stable Diffusion 3 paper.

The paper shows that creating 8 equidistant timesteps and calculating the average loss on them will result in a highly correlated loss to external validation methods such as CLIP or FID score.

This pr's implementation rather than creating 8 stratified timesteps per sample, only applies one of these equidistant timesteps to each sample in a round-robin fashion. Aggregated over many samples in a validation set, this should give a similar validation score as the full timestep method, but will process more validation samples quickly.

### Implementations
- Integrates the image generation evaluation in the validation step, users can 
- Refactors and combines eval job_config with validation
    - Adds an `all_timesteps` option to the job_config to choose whether to use round robin timesteps or full timesteps per sample
- Creates validator class and validation dataloader for flux, validator dataloader handles generating timesteps for round-robin method of validation

### Enabling all timesteps
Developers can enable the full timestamp method of validation by setting `all_timesteps = True` in the flux validation job config. Enabling all_timesteps may require tweaking some hyperparams `validation.local_batch_size, validation.steps` to prevent spiking memory and optimizing throughput. By using a ratio of around 1/4 for `validation.local_batch_size` to `training.local_batch_size` will not spike the memory higher than training when `fsdp = 8`.

Below we can see the difference between round robin and all timesteps. In the comparison the total number of validation samples processed is the same, but in `all_timesteps=True` configuration we have to lower the batch size to prevent memory spiking. All timesteps also achieves a higher throughput (tps) but still processes total samples of validation set more slowly.

| Round Robin (batch_size=32, steps=1, fsdp=8) | All Timesteps (batch_size=8, steps=4, fsdp=8) |
| ---- | --- |
| Screenshot 2025-08-01 at 3 46 42 PM | Screenshot 2025-08-01 at 3 30 10 PM |

### CODE REVIEW

File: torchtitan/experiments/flux/dataset/flux_dataset.py [hunk 1]
```diff
@@ -155,6 +193,7 @@ class FluxDataset(IterableDataset, Stateful):
     dp_rank (int): Data parallel rank.
     dp_world_size (int): Data parallel world size.
     infinite (bool): Whether to loop over the dataset infinitely.
+    generate_timesteps (booL): Generate stratified timesteps in round-robin style for validation
```
Comment (tianyu-l) [inline torchtitan/experiments/flux/dataset/flux_dataset.py]:
```suggestion
    generate_timesteps (bool): Generate stratified timesteps in round-robin style for validation
```

File: torchtitan/experiments/flux/dataset/flux_dataset.py [hunk 2]
```diff
@@ -242,8 +288,10 @@ def __iter__(self):
 
             # skip low quality image or image with color channel = 1
             if sample_dict["image"] is None:
+                # sample_id = sample.get('sample_id')
```
Comment (tianyu-l) [inline torchtitan/experiments/flux/dataset/flux_dataset.py]: remove comment

File: torchtitan/experiments/flux/dataset/flux_dataset.py (hunk 1 above)
Comment (tianyu-l) [reply torchtitan/experiments/flux/dataset/flux_dataset.py]: let's try to create a subclass to do this, as discussed offline.
o/w it's easy to confuse with timestep concept during training / inference.

File: torchtitan/experiments/flux/job_config.py [hunk 3]
```diff
@@ -45,6 +47,8 @@ class Eval:
     """How many denoising steps to sample when generating an image"""
     eval_freq: int = 100
     """Frequency of evaluation/sampling during training"""
+    save_imgs: int = 1
+    """ How many images to generate and save in validation, -1 means same number as steps"""
```
Comment (tianyu-l) [inline torchtitan/experiments/flux/job_config.py]: explain that the source prompt is coming from validation dataset and taken from "the beginning"

File: torchtitan/experiments/flux/job_config.py [hunk 4]
```diff
@@ -45,6 +47,8 @@ class Eval:
     """How many denoising steps to sample when generating an image"""
     eval_freq: int = 100
     """Frequency of evaluation/sampling during training"""
+    save_imgs: int = 1
```
Comment (tianyu-l) [inline torchtitan/experiments/flux/job_config.py]: call it something like `save_imgs_count`? `save_imgs` sounds like a bool

File: torchtitan/experiments/flux/job_config.py [hunk 5]
```diff
@@ -36,7 +38,7 @@ class Encoder:
 
 
 @dataclass
-class Eval:
+class Validation(Validation):
```
Comment (tianyu-l) [inline torchtitan/experiments/flux/job_config.py]: do we have to inherit? `Training` is not inheriting.

File: torchtitan/experiments/flux/model/validate.py [hunk 6]
```diff
@@ -0,0 +1,278 @@
...
+                # Create positional encodings
+                latent_pos_enc = create_position_encoding_for_latents(
+                    bsz, latent_height, latent_width, POSITION_DIM
+                )
+                text_pos_enc = torch.zeros(bsz, t5_encodings.shape[1], POSITION_DIM)
+
+                # Patchify: Convert latent into a sequence of patches
+                latents = pack_latents(latents)
+
+                latent_noise_pred = model(
```
Comment (tianyu-l) [inline torchtitan/experiments/flux/model/validate.py]: please adapt to https://github.com/pytorch/torchtitan/pull/1494 for the amp part

File: torchtitan/experiments/flux/model/validate.py [hunk 7]
```diff
@@ -0,0 +1,278 @@
...
+                batch=input_dict,
+            )
+            labels = input_dict["img_encodings"].to(device_type)
+            clip_encodings = input_dict["clip_encodings"]
+            t5_encodings = input_dict["t5_encodings"]
+
+            bsz = labels.shape[0]
+
+            # To evaluate all 8 timesteps per sample do
+            if self.all_timesteps:
```
Comment (tianyu-l) [inline torchtitan/experiments/flux/model/validate.py]: I think it's OK to keep this.

File: torchtitan/experiments/flux/model/validate.py [hunk 8]
```diff
@@ -0,0 +1,278 @@
...
+        self.metrics_processor.log_validation(loss=global_avg_loss, step=step)
+
+        # Calculate and log validation timing
+        validation_end_time = time.time()
+        validation_duration = validation_end_time - validation_start_time
+
+        # Log timing information
+        from torchtitan.tools.logging import logger
+
+        logger.info(f"Validation step {step} completed in {validation_duration:.3f}s ")
```
Comment (tianyu-l) [inline torchtitan/experiments/flux/model/validate.py]: is this for debugging? we didn't need this in llama3 validator

File: torchtitan/experiments/flux/model/validate.py [hunk 9]
```diff
@@ -0,0 +1,278 @@
...
+                self.job_config.validation.steps != -1
+                and num_steps >= self.job_config.validation.steps
+            ):
+                break
+
+            prompt = input_dict.pop("prompt")
+
+            save_imgs = self.job_config.validation.save_imgs
+            if save_imgs == -1 or num_steps < save_imgs:
+                t5_tokenizer, clip_tokenizer = build_flux_tokenizer(self.job_config)
```
Comment (tianyu-l) [inline torchtitan/experiments/flux/model/validate.py]: why build this multiple times?

File: torchtitan/experiments/flux/dataset/flux_dataset.py [hunk 10]
```diff
@@ -103,6 +104,38 @@ def _cc12m_wds_data_processor(
...
+        sample: A sample from dataset
+        t5_encoder: T5 encoder
+        clip_encoder: CLIP encoder
+        output_size: The output image size
+
+    """
+    img = _process_cc12m_image(sample["image"], output_size=output_size)
+    prompt = sample["caption"]
+    if isinstance(prompt, list):
+        prompt = " ".join(prompt)
```
Comment (tianyu-l) [inline torchtitan/experiments/flux/dataset/flux_dataset.py]: in validate.py you are taking `prompt[0]`. Do you still need to concatenate here?

File: torchtitan/experiments/flux/job_config.py (hunk 4 above)
Comment (tianyu-l) [reply torchtitan/experiments/flux/job_config.py]: also, can you show examples of generate img? Just to verify capability

File: torchtitan/experiments/flux/dataset/flux_dataset.py (hunk 10 above)
Comment (wesleytruong) [reply torchtitan/experiments/flux/dataset/flux_dataset.py]: This concatenation is because coco dataset gives the prompts in a list of strings instead of a single string for one sample. This part just combines the list-style prompt into a single string.

The `prompt[0]` in validation is since we only generate an image from the first sample in each validation batch
Edit: I'm changing this so that it will generate the correct number of images regardless of batch size

**Comment by wesleytruong** [comment]
@tianyu-l addressed comments.

- added all_timesteps option to job_config,
- separated flux dataset and flux validation dataset to separate the timestep generation logic that is only used for validation.
-  changed save_img_count to generate images more logically rather than one per batch

File: torchtitan/experiments/flux/job_config.py (hunk 4 above)
Comment (wesleytruong) [reply torchtitan/experiments/flux/job_config.py]: > also, can you show examples of generate img? Just to verify capability

Here's a generated image to show that generation works. The image isn't good currently since I don't have pre-trained weights but will revisit after adding hf conversion
image

File: torchtitan/experiments/flux/__init__.py [hunk 11]
```diff
@@ -17,6 +17,7 @@
 from .model.args import FluxModelArgs
 from .model.autoencoder import AutoEncoderParams
 from .model.model import FluxModel
+from .model.validate import build_flux_validator
```
Comment (tianyu-l) [inline torchtitan/experiments/flux/__init__.py]: `validate` doesn't sound part of `model`. Let's just put it under `flux/` root folder

File: torchtitan/experiments/flux/dataset/flux_dataset.py (hunk 10 above)
Comment (tianyu-l) [inline torchtitan/experiments/flux/dataset/flux_dataset.py]: If you look at one example
```
[
"A desktop computer monitor sitting next to a keyboard.",
"A computer, keyboard, modem, and mouse are sitting on a work desk.",
"a computer monitor, keyboard, and mouse sit on a table.",
"A black desktop computer atop a wooden table.",
"a desktop computer monitor with a keyboard and mouse"
]
It's always a list of alternative captions. I think we should only pick one (maybe the first one for deterministic training / eval).
```

File: torchtitan/experiments/flux/dataset/flux_dataset.py [hunk 12]
```diff
@@ -242,8 +280,10 @@ def __iter__(self):
 
             # skip low quality image or image with color channel = 1
             if sample_dict["image"] is None:
+                # sample_id = sample.get('sample_id')
```
Comment (tianyu-l) [inline torchtitan/experiments/flux/dataset/flux_dataset.py]: let's remove comment

File: torchtitan/experiments/flux/model/validate.py [hunk 13]
```diff
@@ -0,0 +1,275 @@
...
+
+        for input_dict, labels in self.validation_dataloader:
+            if (
+                self.job_config.validation.steps != -1
+                and num_steps >= self.job_config.validation.steps
+            ):
+                break
+
+            prompt = input_dict.pop("prompt")
+            if not isinstance(prompt, list):
```
Comment (tianyu-l) [inline torchtitan/experiments/flux/model/validate.py]: To educate me: if `prompt` is yielded from the dataset as a `str`, would the dataloader batchify it into a `list`? Could you please share source on this behavior?

**Review by tianyu-l — COMMENTED** [review]
Looks great! Had some final comments.

File: torchtitan/experiments/flux/model/validate.py (hunk 13 above)
Comment (wesleytruong) [reply torchtitan/experiments/flux/model/validate.py]: The torch dataloader handles this batchifying all the way up the inheritance stack `ParallelAwareDataloader -> StatefulDataLoader -> DataLoader`. And the `FluxDataset` gets passed to the `ParallelAwareDataLoader` in the `build_dataloader_fn`. I'm not sure of the exact implementation details and edge cases but the api wants for an IterableDataset with an overloaded __iter__ for datastream. https://docs.pytorch.org/docs/stable/data.html#torch.utils.data.IterableDataset
